### DISCUSSION TITLE
EAP-MS-CHAPv2 verification failed

### DISCUSSION THREAD

**Discussion by tonus01** [strongswan/strongswan#1367] [General]
Hello. Please help. I using 100%  similar config from manual:
https://www.vultr.com/docs/install-a-strongswan-ipsec-vpn-server-on-centos-7/
I i use the same config:
```
config setup
        strictcrlpolicy=yes
        uniqueids=never

conn ikev2
        auto=add
        keyexchange=ikev2
        forceencaps=yes
        ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024!
        esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!
        dpdaction=clear
        dpddelay=300s
        rekey=no
        left=%any
        leftid=@vpn.mydomain.top
        leftcert=fullchain.pem
        leftsendcert=always
        leftsubnet=0.0.0.0/0
        right=%any
        rightid=%any
        rightauth=eap-mschapv2
        rightdns=8.8.8.8,8.8.4.4
        rightsourceip=10.8.0.0/16
        rightsendcert=never
        eap_identity=%identity
```
Log details(not full, only important):
```
05[IKE] authentication of 'vpn.mydomain.top' (myself) with RSA signature successful
...
07[ENC] parsed IKE_AUTH request 2 [ EAP/RES/ID ]
07[IKE] received EAP identity 'user'
07[IKE] initiating EAP_MSCHAPV2 method (id 0x81)
07[ENC] generating IKE_AUTH response 2 [ EAP/REQ/MSCHAPV2 ]
...
08[ENC] parsed IKE_AUTH request 3 [ EAP/RES/MSCHAPV2 ]
08[IKE] EAP-MS-CHAPv2 verification failed, retry (1)
10[MGR] ignoring request with ID 3, already processing
```
When i try to connect, Windows 10 ask password for my user, but password already was in vpn connection config. If i write correct password Windows show error "can't connect".

**Comment by tonus01** [comment]
I find problem with ipsec.secrets
Connection work if use password only with digits , example "123456"
When i use letters, for example "abc123" connectiob not work and i recieve EAP-MS-CHAPv2 verification failed
How i can solve it?

**Reply by tobiasbrunner** [reply]
Could this be a problem with the keyboard layout on the client (i.e. the pressed keys don't correspond to the expected characters)? Or are you using non-ASCII characters in the password?

**Reply by zakyhassani** [reply]
Yeah this is an almost a year thread, but I recently have a very similar issue. The difference is I tried not from Windows. 
I tried to create an IPSec server on Ubuntu 22.04 with very similar config, with EAP-MS-CHAPv2 too. I tried to connect from Mikrotik to the server, the result was the same error: EAP-MS-CHAPv2 verification failed. I tried to change the password to only digits, it works. It can't work if I use alphanumeric password.

**Reply by tobiasbrunner** [reply]
EAP-MSCHAPv2 requires that the password is encoded in UTF-16LE before it is hashed. Depending on the characters in the password, its original character encoding and any conversions the implementation performs, that might not yield the same value on both ends. Note that strongSwan assumes the password (as read from the config, database, or wherever else it comes from) is encoded in UTF-8, which it then converts to UTF-16LE.

**Reply by zakyhassani** [reply]
I see, thanks for the info. AFAIK Winbox by default doesn't necessarily use Unicode. I think it uses Windows ANSI. So, by your explanation, Windows ANSI and UTF-8 seem to produce different binary when converted to UTF-16LE. I'll try to set Mikrotik to use a UTF-8 encoded string as the password then.

edit: it works. thanks.

**Reply by tobiasbrunner** [reply]
Windows will definitely also use Unicode to encode the passwords (see [RFC 2759](https://datatracker.ietf.org/doc/html/rfc2759#section-4)).

**Reply by zakyhassani** [reply]
Hmmmmmm. It works now after I manually add UTF-8 binary encoded string in Mikrotik's terminal and created the identity with that string as the password. (UTF-8 I inserted, for example, is something like this: "\74\65\73\74\68\65\72\65")
I don't know what the explanation is, but as long as it works, I'll take it.

**Reply by tobiasbrunner** [reply]
As I wrote above, it depends on the original encoding and what an implementation does with it afterwards. Could very well be that there is an issue on the MikroTik side (e.g. with the conversion from the encoding in the terminal/console to the storage or later during the EAP-MSCHAPv2 process). However, the password there is simple ASCII ("testhere", which has the same encoding in UTF-8), so it's definitely weird that there would be an issue with it.

**Reply by zakyhassani** [reply]
Ofc it's "testhere", it's only the example of how I put it in Mikrotik, I won't share my own password online for obvious reason.
Yeah it's really weird. I set it back to directly input my password in the winbox UI, and the error come back. When I changed back to that ASCII/UTF-8  just like this "testhere": "\74\65\73\74\68\65\72\65", the error was gone again.

**Reply by tobiasbrunner** [reply]
> Ofc it's "testhere", it's only the example of how I put it in Mikrotik

Obviously, but you made it sound like even that example won't work unless you encode it explicitly (you still do). If it was only a problem if there actually were non-ASCII characters in the password, then it would make a lot more sense. But not for simple ASCII passwords.

>  I set it back to directly input my password in the winbox UI

Oh, I wasn't aware Winbox is actually a MicroTik thing, I thought you used that as a short form of "Windows box" above :)

Anyway, it's definitely possible that UI does something weird with the entered password. Maybe depending on the system it runs on or some setting there, because if this always had issues with simple ASCII passwords I feel somebody would have noticed this before.
